pkg/services/oauthserver/external_service.go

```diff
@@ -0,0 +1,161 @@
+package oauthserver
+
+import (
+	"context"
+	"strconv"
+	"strings"
+
+	"github.com/ory/fosite"
+
+	ac "github.com/grafana/grafana/pkg/services/accesscontrol"
+	"github.com/grafana/grafana/pkg/services/user"
+)
+
+type KeyResult struct {
+	URL        string `json:"url,omitempty"`
+	PrivatePem string `json:"private,omitempty"`
+	PublicPem  string `json:"public,omitempty"`
+	Generated  bool   `json:"generated,omitempty"`
+}
+
+type ExternalServiceDTO struct {
+	Name        string     `json:"name"`
+	ID          string     `json:"clientId"`
+	Secret      string     `json:"clientSecret"`
+	RedirectURI string     `json:"redirectUri,omitempty"` // Not used yet (code flow)
+	GrantTypes  string     `json:"grantTypes"`            // CSV value
+	Audiences   string     `json:"audiences"`             // CSV value
+	KeyResult   *KeyResult `json:"key,omitempty"`
+}
+
+type ExternalService struct {
+	ID               int64  `xorm:"id pk autoincr"`
+	Name             string `xorm:"name"`
+	ClientID         string `xorm:"client_id"`
+	Secret           string `xorm:"secret"`
+	RedirectURI      string `xorm:"redirect_uri"` // Not used yet (code flow)
+	GrantTypes       string `xorm:"grant_types"`  // CSV value
+	Audiences        string `xorm:"audiences"`    // CSV value
+	PublicPem        []byte `xorm:"public_pem"`
+	ServiceAccountID int64  `xorm:"service_account_id"`
+	// SelfPermissions are the registered service account permissions (registered and managed permissions)
+	SelfPermissions []ac.Permission
+	// ImpersonatePermissions is the restriction set of permissions while impersonating
+	ImpersonatePermissions []ac.Permission
+
+	// SignedInUser refers to the current Service Account identity/user
+	SignedInUser      *user.SignedInUser
+	Scopes            []string
+	ImpersonateScopes []string
+}
+
+func (c *ExternalService) ToDTO() *ExternalServiceDTO {
+	c2 := ExternalServiceDTO{
+		Name:        c.Name,
+		ID:          c.ClientID,
+		Secret:      c.Secret,
+		GrantTypes:  c.GrantTypes,
+		Audiences:   c.Audiences,
+		RedirectURI: c.RedirectURI,
+	}
+	if len(c.PublicPem) > 0 {
+		c2.KeyResult = &KeyResult{PublicPem: string(c.PublicPem)}
+	}
+	return &c2
+}
+
+func (c *ExternalService) LogID() string {
+	return "{name: " + c.Name + ", clientID: " + c.ClientID + "}"
```

Comment:
Nit: Would it be possible to log the Name and ClientID separately? Currently in the log it would look like this:
`client={name: someextsvc, clientID: 123456csafd}`. And I guess it's more difficult, slower to index/search than having `name=someextsvc` and `clientid=12345csafd` in the log.